### PR TITLE
[model] feat: add Qwen3ForSequenceClassification to support sequence embedding classification

### PR DESCRIPTION
**Pull request by yiwzhao**
[model] feat: add Qwen3ForSequenceClassification to support sequence embedding classification

### CODE REVIEW

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 1]
```diff
@@ -725,10 +727,95 @@ def forward(
...
+    - The current loss function directly uses single-label CrossEntropyLoss.
+    """
+
+    base_model_prefix = "model"
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.num_labels = config.num_labels
+        # we're not using AutoModel here, but our own Qwen3Model.
+        setattr(self, self.base_model_prefix, Qwen3Model(config))
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3/modeling_qwen3.py]: I don't quite understand this? can it just be?        self.model = Qwen3Model(config)

like https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3/modeling_qwen3.py#L464C1-L465C1

Comment (yiwzhao) [reply veomni/models/transformers/qwen3/modeling_qwen3.py]: Yes, in this class they are equal.

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 2]
```diff
@@ -725,10 +727,94 @@ def forward(
         )
 
 
+class Qwen3ForSequenceClassification(Qwen3PreTrainedModel):
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3/modeling_qwen3.py:730]: nit: add modification note.

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 3]
```diff
@@ -725,10 +727,94 @@ def forward(
...
+        input_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Cache] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        cu_seqlens: Optional[torch.Tensor] = None,
+        max_seqlen: Optional[int] = None,
+        **kwargs: Unpack[FlashAttentionKwargs],
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3/modeling_qwen3.py:763]: check if it conflicts with the unpack FlashAttentionKwargs

File: veomni/ops/loss.py [hunk 4]
```diff
@@ -90,3 +91,23 @@ def causallm_loss_function(
...
+) -> torch.Tensor:
+    # local sum loss
+    # CrossEntropyLoss(reduction="none") + mask + sum
+    per = loss_fct(logits, labels)  # [N] if reduction="none"
+    valid = labels != ignore_index
+    loss_sum = (per * valid).sum()
+    cnt = valid.sum().to(dtype=loss_sum.dtype)
+
+    if sp_group is not None:
+        dist.all_reduce(loss_sum, op=dist.ReduceOp.SUM, group=sp_group)
```
Comment (piyifan123) [inline veomni/ops/loss.py:111]: see if we can reuse reduce_sequence_parallel_loss

File: veomni/ops/loss.py [hunk 5]
```diff
@@ -90,3 +91,23 @@ def causallm_loss_function(
         loss = reduce_sequence_parallel_loss(loss, num_valid_tokens)
 
     return loss, logits
+
+
+def seqcls_token_loss_sp_aware(
```
Comment (piyifan123) [inline veomni/ops/loss.py:96]: add docstring and unit tests.

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 6]
```diff
@@ -725,10 +727,94 @@ def forward(
...
+            **kwargs,
+        )
+
+        hidden_states = transformer_outputs.last_hidden_state
+        logits = self.score(hidden_states)
+
+        loss = None
+        if labels is not None:
+            # labels are token-level now, shape must match logits tokens
+            if logits.dim() == 3:
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3/modeling_qwen3.py:784]: can we assume always we have B?

File: veomni/ops/loss.py (hunk 5 above)
Comment (Coach257) [inline veomni/ops/loss.py:96]: why not use the `causallm_loss_function`

Comment (piyifan123) [reply veomni/ops/loss.py:96]: Shift label is not needed for seq classification task in general. And it's more readable and less error prone to have a separate loss function for classification from ntp loss imo

File: veomni/models/transformers/qwen3/modeling_qwen3.py (hunk 3 above)
Comment (yiwzhao) [reply veomni/models/transformers/qwen3/modeling_qwen3.py:763]: There won't be conflict because the key names are different. FlashAttentionKwargs only contains: cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k. My key names here are use_cache, cu_seqlens, and max_seqlen.

File: veomni/models/transformers/qwen3/modeling_qwen3.py (hunk 6 above)
Comment (yiwzhao) [reply veomni/models/transformers/qwen3/modeling_qwen3.py:784]: Yes, we can. I‘ll remove the branch where logits.dim() == 2.

File: veomni/models/transformers/qwen3/modeling_qwen3.py (hunk 2 above)
Comment (yiwzhao) [reply veomni/models/transformers/qwen3/modeling_qwen3.py:730]: Added

File: veomni/ops/loss.py (hunk 5 above)
Comment (yiwzhao) [reply veomni/ops/loss.py:96]: Added.

File: veomni/ops/loss.py (hunk 4 above)
Comment (yiwzhao) [reply veomni/ops/loss.py:111]: Yes, we can. I implemented it in the new MR: https://github.com/ByteDance-Seed/VeOmni/pull/322

File: veomni/models/transformers/qwen3/modeling_qwen3.py (hunk 1 above)
Comment (yiwzhao) [reply veomni/models/transformers/qwen3/modeling_qwen3.py]: Yes, in this scenario they are the same.

File: veomni/models/transformers/qwen3/modeling_qwen3.py (hunk 3 above)
Comment (piyifan123) [reply veomni/models/transformers/qwen3/modeling_qwen3.py:763]: why not use FlashAttentionKwargs instead? i don't see the reason why you need to define your own cu_seqlens, and max_seqlen which could cause HF transformers fa_util does not read the proper FA kwargs and cause cpu sync calculation.

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 7]
```diff
@@ -725,10 +727,94 @@ def forward(
...
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Cache] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        cu_seqlens: Optional[torch.Tensor] = None,
+        max_seqlen: Optional[int] = None,
+        **kwargs: Unpack[FlashAttentionKwargs],
+    ) -> SequenceClassifierOutputWithPast:
+        transformer: Qwen3Model = getattr(self, self.base_model_prefix)
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3/modeling_qwen3.py:765]: why can't we just simply write self.model here?

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 8]
```diff
@@ -725,10 +727,98 @@ def forward(
...
+            input_ids,
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            cu_seqlens=cu_seqlens,
+            max_seqlen=max_seqlen,
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (piyifan123) [reply veomni/models/transformers/qwen3/modeling_qwen3.py:776]: yes, same issue here. why define your own cu_seqlens and max_seqlen while the underlying model does not even understand it...

File: veomni/models/transformers/qwen3/modeling_qwen3.py [hunk 9]
```diff
@@ -725,10 +727,94 @@ def forward(
...
+
+        hidden_states = transformer_outputs.last_hidden_state
+        logits = self.score(hidden_states)
+
+        loss = None
+        if labels is not None:
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3/modeling_qwen3.py:782]: i still think it's too complex while we can basically go with the linearcrossentropy path as the NTP head does by making sure that labels are set with proper ignore_index there.

File: veomni/ops/loss.py (hunk 5 above)
Comment (piyifan123) [inline veomni/ops/loss.py:96]: i think you should have the exact same input format as causallm_loss_function, i.e., the part I don't think makes a lot of sense is: the current definition requires model to calculate logits which will cause 2 problems:
1. hidden_states -> logits projection logic is leaked to the modeling code leads to more complexity. 
2. lose the chance to use a fusedlinearcrossentropy kernel for performance.
3. introduce a different pattern from ntp loss while they can be pretty much very similar for people to understand.
